twoliter/embedded/img2img

```diff
@@ -0,0 +1,288 @@
+if [[ ${NEW_ROOT_SIZE} -gt ${OLD_ROOT_SIZE} ]]; then
+  echo "new root image size of ${NEW_ROOT_SIZE} bytes is greater than old size of ${OLD_ROOT_SIZE} bytes" >&2
+  exit 1
+fi
+dd if="${ROOT_IMAGE}" of="${OS_IMAGE}" conv=notrunc seek="${ROOT_START}"
+echo "Wrote files, recomputing verity"
+
+# step 5: regenerate the verity image
+VERITY_IMAGE="${WORKDIR}/root.verity"
+VERITY_START="$(sgdisk --print "${OS_IMAGE}" | awk '/BOTTLEROCKET-HASH-A/{ print $2 }')"
+VERITY_END="$(sgdisk --print "${OS_IMAGE}" | awk '/BOTTLEROCKET-HASH-A/{ print $3 }')"
+VERITY_SIZE="$((VERITY_END - VERITY_START + 1))"
+veritysetup_output="$(veritysetup_format "${ROOT_IMAGE}" "${VERITY_IMAGE}")"
+VERITY_DATA_4K_BLOCKS="$(grep '^Data blocks:' <<<"${veritysetup_output}" | awk '{ print $NF }')"
+VERITY_DATA_512B_BLOCKS="$((VERITY_DATA_4K_BLOCKS * 8))"
+VERITY_ROOT_HASH="$(grep '^Root hash:' <<<"${veritysetup_output}" | awk '{ print $NF }')"
+VERITY_SALT="$(grep '^Salt:' <<<"${veritysetup_output}" | awk '{ print $NF }')"
+OLD_VERITY_SIZE="$((VERITY_SIZE * 512))"
+NEW_VERITY_SIZE="$(stat --format='%s' "${VERITY_IMAGE}")"
+if [[ ${NEW_VERITY_SIZE} -gt ${OLD_VERITY_SIZE} ]]; then
+  echo "new hash image size of ${NEW_VERITY_SIZE} bytes is greater than old size of ${OLD_VERITY_SIZE} bytes" >&2
+  exit 1
+fi
+
+# step 6: write the verity image back
+dd if="${VERITY_IMAGE}" of="${OS_IMAGE}" conv=notrunc seek="${VERITY_START}"
+
+# step 7: extract the boot filesystem
+BOOT_IMAGE="${WORKDIR}/boot.ext4"
+BOOT_START="$(sgdisk --print "${OS_IMAGE}" | awk '/BOTTLEROCKET-BOOT-A/{ print $2 }')"
+BOOT_END="$(sgdisk --print "${OS_IMAGE}" | awk '/BOTTLEROCKET-BOOT-A/{ print $3 }')"
+BOOT_SIZE="$((BOOT_END - BOOT_START + 1))"
+dd if="${OS_IMAGE}" of="${BOOT_IMAGE}" count="${BOOT_SIZE}" skip="${BOOT_START}"
+BOOT_MOUNT="${WORKDIR}/boot"
+mkdir -p "${BOOT_MOUNT}"
+debugfs -R "rdump / ${BOOT_MOUNT}" "${BOOT_IMAGE}"
+
+# step 8: maybe secure boot
+if [[ "${UEFI_SECURE_BOOT}" == "yes" ]]; then
+  EFI_IMAGE="${WORKDIR}/efi.vfat"
+  EFI_START="$(sgdisk --print "${OS_IMAGE}" | awk '/EFI-SYSTEM/{ print $2 }')"
+  EFI_END="$(sgdisk --print "${OS_IMAGE}" | awk '/EFI-SYSTEM/{ print $3 }')"
+  EFI_SIZE="$((EFI_END - EFI_START + 1))"
+  dd if="${OS_IMAGE}" of="${EFI_IMAGE}" count="${EFI_SIZE}" skip="${EFI_START}"
+  EFI_MOUNT="${WORKDIR}/efi"
+  mkdir -p "${EFI_MOUNT}/EFI/BOOT"
+  mcopy -i "${EFI_IMAGE}" -sv ::/EFI/BOOT "${EFI_MOUNT}/EFI/"
+
+  # Do the setup required for `pesign` and `gpg` signing and
+  # verification to "just work" later on, regardless of which
+  # type of signing profile we have.
+  declare -a shim_sign_key code_sign_key
+  sbsetup shim_sign_key code_sign_key
+
+  # remove the original signing certificate to avoid confusion
+  rm "${EFI_MOUNT}/EFI/BOOT/db.cer" "${EFI_MOUNT}/EFI/BOOT/db.crt"
+
+  # replace artifacts in efi filesystem
+  pushd "${EFI_MOUNT}/EFI/BOOT" >/dev/null
+  replace_vendor_certs shim_sign_key code_sign_key "${UEFI_SECURE_BOOT}"
```

Comment:
This smells like an error - we need to call this even if we are not doing Secure Boot for consistency with `rpm2img`, because this is what does the no-op replacements in shim and grub. But because we know it's a Secure Boot image, we know that doesn't apply to the artifacts we've extracted.

I'd recommend splitting the function into two, like `replace_vendor_certs` and `replace_dummy_certs`, and then `rpm2img` can call one or the other depending on whether Secure Boot is active.

Reply:
In the case that UEFI_SECURE_BOOT is "no" this block never gets called and we don't modify EFI at all. My hope is we only touch the bare minimum when repacking. If we do want to touch both though, I wonder if we should just generate a fresh EFI image each time.